### PR TITLE
Adding the global group hash methods and cache for cluster syncronization

### PR DESCRIPTION
**Pull request by pereyra-m**
|Related issue|
|---|
|#11793|

## Description

This PR creates the `wdb_get_global_group_hash()` method to calculate the hash of the whole `group_local_hash` column in the **agent** table. The global group hash is stored in cache to avoid calculating it every time and this cache is cleared every time a new group assignment is done.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation

- Memory tests for Linux
  - [x] Scan-build report

### CODE REVIEW

**Comment by pereyra-m** [comment]
### Scan-build

The tool doesn't report any error

scan_build

### Manual tests

Some manual calls to `wdb_get_global_group_hash()` were introduced to test the mechanism.
These are the agents registered for the test

group_pre_state

The first time, the value is calculated and stored. The value is the same than the one returned by the `sha1sum` terminal command

sha_calculated

We can see in the logs that the value is stored

value_stored

Then, when the method is called a second time, the value is read from the cache

cache_found

It was also confirmed that the method `wdb_global_set_agent_groups()` clears the cache and that the value has to be generated again.

File: src/wazuh_db/wdb_integrity.c [hunk 1]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
     cJSON_Delete(j_sync_info);
     return result;
 }
+
+int wdb_get_global_group_hash(wdb_t * wdb, os_sha1 hexdigest) {
+    if (OS_SUCCESS == wdb_global_group_hash_cache_operations("read", hexdigest)) {
+        mdebug2("A global group hash was found in cache");
```
Comment (palaciosjeremias) [inline src/wazuh_db/wdb_integrity.c:645]: We can rephrase this like
`Using group hash from cache`

File: src/wazuh_db/wdb_integrity.c [hunk 2]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
+            return OS_INVALID;
+        }
+
+        sqlite3_stmt* stmt = wdb_init_stmt_in_cache(wdb, WDB_STMT_GLOBAL_GROUP_HASH_GET);
+        if (!stmt) {
+            return OS_INVALID;
+        }
+
+        int result = wdb_calculate_stmt_checksum(wdb, stmt, WDB_GENERIC_COMPONENT, hexdigest);
+        if(result == 1) {
```
Comment (palaciosjeremias) [inline src/wazuh_db/wdb_integrity.c:659]: We can simplify this as:
`if (wdb_calculate_stmt_checksum(wdb, stmt, WDB_GENERIC_COMPONENT, hexdigest)) {`

File: src/wazuh_db/wdb_integrity.c [hunk 3]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
+        return OS_SUCCESS;
+    } else {
+        if(!wdb) {
+            mdebug1("Missing DB pointer to calculate the global group hash.");
+            return OS_INVALID;
+        }
+
+        sqlite3_stmt* stmt = wdb_init_stmt_in_cache(wdb, WDB_STMT_GLOBAL_GROUP_HASH_GET);
+        if (!stmt) {
+            return OS_INVALID;
```
Comment (palaciosjeremias) [inline src/wazuh_db/wdb_integrity.c:655]: Shouldn´t we have another debug message here like in `if(!wdb) {`?

**Review by palaciosjeremias — CHANGES_REQUESTED** [review]

File: src/wazuh_db/wdb.c [hunk 4]
```diff
@@ -174,6 +174,7 @@ static const char *SQL_STMT[] = {
     [WDB_STMT_GLOBAL_GROUP_SOURCE_SET] = "UPDATE agent SET group_source = ? WHERE id = ?;",
     [WDB_STMT_GLOBAL_GROUP_CSV_GET] = "SELECT `group` from agent where id = ?;",
     [WDB_STMT_GLOBAL_GROUP_CTX_SET] = "UPDATE agent SET 'group' = ?, group_local_hash = ?, group_sync_status = ?, group_source = ? WHERE id = ?;",
+    [WDB_STMT_GLOBAL_GROUP_HASH_GET] = "SELECT group_local_hash FROM agent WHERE group_local_hash IS NOT NULL;",
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb.c:177]: Done!

File: src/wazuh_db/wdb.h [hunk 5]
```diff
@@ -1488,6 +1490,28 @@ void wdbi_set_last_completion(wdb_t * wdb, wdb_component_t component, long times
 
 int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component);
 
+/**
+ * @brief Method to obtain the hash of the whole group_local_hash column in agent table.
+ *        If the cache is empty, the global group hash is calculated and stored.
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb.h:1495]: Done!

File: src/wazuh_db/wdb.h [hunk 6]
```diff
@@ -1488,6 +1490,28 @@ void wdbi_set_last_completion(wdb_t * wdb, wdb_component_t component, long times
...
+ * @param wdb The DB pointer structure.
+ * @param hexdigest Variable to return the global group hash.
+ * @return int OS_SUCCESS if the hexdigest variable was written with the global group hash value, OS_INVALID otherwise.
+ */
+int wdb_get_global_group_hash(wdb_t * wdb, os_sha1 hexdigest);
+
+/**
+ * @brief Method to perform all the required operations over the global group hash cache.
+ *
+ * @param mode      "read" : OS_INVALID if there is no value in cache. OS_SUCCESS if a value was found and stored in hexdigest
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb.h:1507]: Done!

File: src/wazuh_db/wdb.h [hunk 7]
```diff
@@ -357,6 +358,7 @@ typedef enum {
     WDB_SYSCOLLECTOR_NETINFO,        ///< Net info integrity monitoring.
     WDB_SYSCOLLECTOR_HWINFO,         ///< Hardware info integrity monitoring.
     WDB_SYSCOLLECTOR_OSINFO,         ///< OS info integrity monitoring.
+    WDB_GENERIC_COMPONENT,           ///< Defined to re-utilize the methods outside the integrity library
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb.h:361]: Done!

File: src/wazuh_db/wdb_integrity.c [hunk 8]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
     return result;
 }
+
+int wdb_get_global_group_hash(wdb_t * wdb, os_sha1 hexdigest) {
+    if (OS_SUCCESS == wdb_global_group_hash_cache_operations("read", hexdigest)) {
+        mdebug2("A global group hash was found in cache");
+        return OS_SUCCESS;
+    } else {
+        if(!wdb) {
+            mdebug1("Missing DB pointer to calculate the global group hash.");
```
Comment (DProvinciani) [inline src/wazuh_db/wdb_integrity.c:649]: I would rephrase this with something like `Database structure not initialized. Unable to calculate global group hash`

File: src/wazuh_db/wdb_integrity.c [hunk 9]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
+        }
+
+        int result = wdb_calculate_stmt_checksum(wdb, stmt, WDB_GENERIC_COMPONENT, hexdigest);
+        if(result == 1) {
+            wdb_global_group_hash_cache_operations("set", hexdigest);
+            mdebug2("New global group hash calculated and stored in cache.");
+            return OS_SUCCESS;
+        } else {
+            mdebug2("No group hash was found to calculate the global group hash.");
+            return OS_INVALID;
```
Comment (DProvinciani) [inline src/wazuh_db/wdb_integrity.c:665]: Just wondering whether this is semantically correct. Cause in fact, the method succeded on its operations but there is no data in the database. In such case, it succeeds but returns empty data. Opinions?

File: src/wazuh_db/wdb.h [hunk 10]
```diff
@@ -1488,6 +1490,28 @@ void wdbi_set_last_completion(wdb_t * wdb, wdb_component_t component, long times
...
+/**
+ * @brief Method to perform all the required operations over the global group hash cache.
+ *
+ * @param mode      "read" : OS_INVALID if there is no value in cache. OS_SUCCESS if a value was found and stored in hexdigest
+ *                  "set"  : OS_SUCCESS after writting the hexdigest value in global_group_hash.
+ *                  "clear": OS_SUCCESS after clearing the global group hash cache.
+ * @param hexdigest Input/Output variable, see "mode".
+ * @return int OS_INVALID in case of an unsupported "mode". See "mode" for the rest of cases.
+ */
+int wdb_global_group_hash_cache_operations(const char* mode, os_sha1 hexdigest);
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb.h:1513]: Done!

[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb.h:1513]: You are right, done!

File: src/wazuh_db/wdb_integrity.c [hunk 11]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
+        }
+    }
+}
+
+int wdb_global_group_hash_cache_operations(const char* mode, os_sha1 hexdigest) {
+    static os_sha1 global_group_hash = {0};
+
+    if (strcmp(mode, "read") == 0) {
+        if (global_group_hash[0] == 0) {
+            return OS_INVALID;
```
Comment (DProvinciani) [inline src/wazuh_db/wdb_integrity.c:675]: Similar to the previous comment here.

File: src/wazuh_db/wdb_integrity.c [hunk 12]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
+        } else {
+            snprintf(hexdigest, sizeof(os_sha1), "%s", global_group_hash);
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:677]: Done!

File: src/wazuh_db/wdb_integrity.c (hunk 1 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:645]: Done!

File: src/wazuh_db/wdb_integrity.c [hunk 13]
```diff
@@ -638,3 +639,52 @@ int wdbi_check_sync_status(wdb_t *wdb, wdb_component_t component) {
...
+            return OS_SUCCESS;
+        }
+    } else if(strcmp(mode, "set") == 0) {
+        snprintf(global_group_hash, sizeof(os_sha1), "%s", hexdigest);
+        return OS_SUCCESS;
+    } else if (strcmp(mode, "clear") == 0) {
+        global_group_hash[0] = 0;
+        return OS_SUCCESS;
+    } else {
+        mdebug1("Invalid mode for global group hash operation.");
```
Comment (DProvinciani) [inline src/wazuh_db/wdb_integrity.c:687]: As the modes are not set/managed by the users, and are specified from the code that we implement, I would log this as `mdebug2`

File: src/wazuh_db/wdb_integrity.c (hunk 2 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:659]: Done!

**Review by DProvinciani — CHANGES_REQUESTED** [review]
Just some minor comments.

File: src/wazuh_db/wdb_integrity.c (hunk 3 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:655]: Here, the `wdb_init_stmt_in_cache()` method will print the corresponding debug message.

File: src/wazuh_db/wdb_integrity.c (hunk 8 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:649]: You are right!

File: src/wazuh_db/wdb_integrity.c (hunk 9 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:665]: It's interesting because the lack of a non-NULL group hash is a valid situation; but I'm returning `OS_INVALID` because the **hexdigest** variable hasn't been written and I didn't want it to be read. What do you think?

File: src/wazuh_db/wdb_integrity.c (hunk 11 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:675]: With this return code, I meant the cache is empty and that the **hexdigest** value hasn't been written. But this isn't a failure condition, you are right

File: src/wazuh_db/wdb_integrity.c (hunk 13 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_integrity.c:687]: Done!

File: src/wazuh_db/wdb_integrity.c (hunk 9 above)
Comment (DProvinciani) [reply src/wazuh_db/wdb_integrity.c:665]: Probably we could keep it as it is implemented right now. The thing is that probably is possible to implement that change that I'm suggesting, but this would involve an extra check in the caller. In this way, we avoid the extra check and simplify the function usage.

File: src/wazuh_db/wdb_integrity.c (hunk 11 above)
Comment (DProvinciani) [reply src/wazuh_db/wdb_integrity.c:675]: Same as before. Let's keep it as it is implemented to avoid an extra check in the caller.
